### PR TITLE
Bjorncs/vespa feed client in vespa hadoop

### PR DESCRIPTION
**Pull request by bjorncs**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: vespa-hadoop/src/main/java/com/yahoo/vespa/hadoop/mapreduce/VespaRecordWriter.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        feeder = createJsonStreamFeeder();
+        initialized = true;
+    }
+
+    private void validateConfig() {
+        if (!config.useSSL()) {
+            throw new IllegalArgumentException("SSL is required for this feed client implementation");
+        }
+        if (config.dryrun()) {
+            throw new IllegalArgumentException("Dryrun is not supported for this feed client implementation");
```
Comment (jonmv) [inline vespa-hadoop/src/main/java/com/yahoo/vespa/hadoop/mapreduce/VespaRecordWriter.java]: We should probably add a dry-run option.

File: vespa-hadoop/src/main/java/com/yahoo/vespa/hadoop/mapreduce/VespaRecordWriter.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        }
+        if (config.dataFormat() != FeedParams.DataFormat.JSON_UTF8) {
+            throw new IllegalArgumentException("Only JSON is support by this feed client implementation");
+        }
+        if (config.proxyHost() != null) {
+            log.warning(String.format("Ignoring proxy config (host='%s', port=%d)", config.proxyHost(), config.proxyPort()));
```
Comment (jonmv) [inline vespa-hadoop/src/main/java/com/yahoo/vespa/hadoop/mapreduce/VespaRecordWriter.java:102]: Hmm, do we need to support this?

File: vespa-hadoop/src/main/java/com/yahoo/vespa/hadoop/mapreduce/VespaRecordWriter.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+    protected VespaRecordWriter(VespaConfiguration config, VespaCounters counters) {
+        this.counters = counters;
+        this.config = config;
+    }
+
+    @Override
+    public void write(Object key, Object data) throws IOException {
+        initializeOnFirstWrite();
+        String json = "[" + data.toString().trim() + "]";
```
Comment (jonmv) [inline vespa-hadoop/src/main/java/com/yahoo/vespa/hadoop/mapreduce/VespaRecordWriter.java:53]: We could probably let the JSON parser accept a series of JSON objects, too, in addition to a JSON array.

Comment (jonmv) [reply vespa-hadoop/src/main/java/com/yahoo/vespa/hadoop/mapreduce/VespaRecordWriter.java:53]: (Alternatively, `SequenceInputStream` can be used here :))

File: vespa-hadoop/src/main/java/com/yahoo/vespa/hadoop/mapreduce/VespaRecordWriter.java [hunk 4]
```diff
@@ -143,4 +154,28 @@ private static int streamsPerConnection(VespaConfiguration config) {
                 .map(hostname -> URI.create(String.format("https://%s:%d/", hostname, config.defaultPort())))
                 .collect(toList());
     }
+
+    private static class DryrunClient implements FeedClient {
```
Comment (jonmv) [inline vespa-hadoop/src/main/java/com/yahoo/vespa/hadoop/mapreduce/VespaRecordWriter.java:158]: Let's move this to the json parser package. I think we should offer some dry-run option for just verifying the JSON, counting documents, etc., there.

**Comment by jonmv** [comment]
Nice :)

File: vespa-hadoop/src/main/java/com/yahoo/vespa/hadoop/mapreduce/VespaRecordWriter.java (hunk 1 above)
Comment (bjorncs) [reply vespa-hadoop/src/main/java/com/yahoo/vespa/hadoop/mapreduce/VespaRecordWriter.java]: I'll add this once we are settled on a http client implementation

File: vespa-hadoop/src/main/java/com/yahoo/vespa/hadoop/mapreduce/VespaRecordWriter.java (hunk 2 above)
Comment (bjorncs) [reply vespa-hadoop/src/main/java/com/yahoo/vespa/hadoop/mapreduce/VespaRecordWriter.java:102]: We are not sure yet - current theory is that it's not necessary. We'll have to try it out in CI.

File: vespa-hadoop/src/main/java/com/yahoo/vespa/hadoop/mapreduce/VespaRecordWriter.java (hunk 1 above)
Comment (bjorncs) [reply vespa-hadoop/src/main/java/com/yahoo/vespa/hadoop/mapreduce/VespaRecordWriter.java]: Added in b525d31

**Comment by bjorncs** [comment]
Replaced by https://github.com/vespa-engine/vespa/pull/18181

**Pull request closed without merging**